### PR TITLE
[WIP] Use symbolic references more liberally.

### PR DESCRIPTION
**Pull request by jckarter**
Introduce symbolic references for indirect references to type context descriptors in other object files, and a new kind for references to protocol conformance descriptors (which will be used to mangle general generic arguments for key paths)

TODO:

- [ ] Update IRGen tests for changed output
- [ ] Fix an issue with reflection of symbolic-referenced protocol contexts (evident in failing `test/Reflection` tests)

Deferred:

- [ ] Handle indirect symbolic references correctly in Reflection tests. This probably means that the Mach-O/ELF/COFF `ObjectMemoryReader`s need to learn to resolve relocations among multiple disk files, alas. For now, I've kept symbolic references across files disabled.

### CODE REVIEW

File: lib/IRGen/IRGenMangler.cpp [hunk 1]
```diff
@@ -171,12 +178,21 @@ mangleSymbolNameForSymbolicMangling(const SymbolicMangling &mangling) {
...
+      = Storage[prefixLen + offset+2]
+      = Storage[prefixLen + offset+3]
+      = Storage[prefixLen + offset+4]
+      = '_';
     Buffer << ' ';
-    appendContext(dc);
+    if (auto ty = referent.dyn_cast<const NominalTypeDecl*>())
+      appendContext(ty);
+    else if (auto proto = referent.dyn_cast<const ProtocolDecl*>())
+      appendContext(proto);
```
Comment (DougGregor) [inline lib/IRGen/IRGenMangler.cpp]: Given that a `ProtocolDecl` is-a `NominalTypeDecl`, we have to be excessively careful with that `referent` type. Do we need the protocol vs. nominal type distinction here?

File: lib/IRGen/Linking.cpp [hunk 2]
```diff
@@ -662,3 +662,247 @@ bool LinkEntity::isAvailableExternally(IRGenModule &IGM) const {
   }
   llvm_unreachable("bad link entity kind");
 }
+
+llvm::Type *LinkEntity::getDefaultDeclarationType(IRGenModule &IGM) const {
```
Comment (DougGregor) [inline lib/IRGen/Linking.cpp]: This is a very nice and mostly unrelated cleanup ;)

File: lib/IRGen/MetadataRequest.cpp [hunk 3]
```diff
@@ -246,22 +247,50 @@ llvm::Constant *IRGenModule::getAddrOfStringForTypeRef(
...
+    } else if (auto cconf =
+                        symbolic.first.dyn_cast<const ProtocolConformance*>()) {
+      // The symbolic reference is to the protocol conformance descriptor of
+      // the referenced conformance.
+      auto conf = const_cast<ProtocolConformance *>(cconf);
+      ref = getAddrOfLLVMVariableOrGOTEquivalent(
+        LinkEntity::forProtocolConformanceDescriptor(conf->getRootNormalConformance()));
+      
+      // \3 - direct reference, \4 - indirect reference
+      baseKind = 5;
```
Comment (DougGregor) [inline lib/IRGen/MetadataRequest.cpp]: Shouldn't this be 3 instead of 5?

File: stdlib/public/runtime/MetadataLookup.cpp [hunk 4]
```diff
@@ -54,17 +54,106 @@ static Demangler getDemanglerForRuntimeTypeResolution() {
...
+    auto conformance = (const ProtocolConformanceDescriptor *)resolvedReference;
+    
+    auto protocolMangling =
+      _buildDemanglingForContext(conformance->getProtocol(), {}, Dem);
+    
+    const ModuleContextDescriptor *module;
+    if (auto retroContext = conformance->getRetroactiveContext()) {
+      module = retroContext->getModuleContext();
+    } else {
+      // FIXME: Assume the protocol's module for a non-retroactive conformance
```
Comment (DougGregor) [inline stdlib/public/runtime/MetadataLookup.cpp]: Non-retroactive conformances could be either in the protocol's module or the conforming type's module. If we need it, we could add a bit to the conformance descriptor to say which one it is...

File: stdlib/public/runtime/MetadataLookup.cpp [hunk 5]
```diff
@@ -653,17 +740,12 @@ _findProtocolDescriptor(const Demangle::NodePointer &node,
...
       (const ContextDescriptor *)symbolicNode->getIndex());
 
   mangledName =
-    Demangle::mangleNode(node,
-                         [&](const void *context) -> NodePointer {
-                           return _buildDemanglingForContext(
-                               (const ContextDescriptor *) context,
-                               {}, Dem);
-                         });
+    Demangle::mangleNode(node, ExpandResolvedSymbolicReferences(Dem));
```
Comment (DougGregor) [inline stdlib/public/runtime/MetadataLookup.cpp]: Thanks for cleaning up the DRY violations here

**Review by DougGregor — COMMENTED** [review]
This looks very promising!

File: docs/ABI/Mangling.rst [hunk 6]
```diff
@@ -42,6 +46,48 @@ mangled name will start with the module name (after the ``_S``).
 In the following, productions which are only _part_ of an operator, are
 named with uppercase letters.
 
+Symbolic references
+~~~~~~~~~~~~~~~~~~~
+
+The Swift compiler emits mangled names into binary images to encode
+references to types for runtime instantiation and reflection.Within an
```
Comment (DougGregor) [inline docs/ABI/Mangling.rst]: Missing space after "reflection."

**Comment by DougGregor** [comment]
Indirect symbolic references will let us re-land https://github.com/apple/swift/pull/19848

File: lib/IRGen/IRGenMangler.cpp (hunk 1 above)
Comment (jckarter) [reply lib/IRGen/IRGenMangler.cpp]: Probably not. I started out giving type and protocols different symbolic reference discriminators, but that ended up being unnecessary.

**Comment by DougGregor** [comment]
Note that this is obsoleted by https://github.com/apple/swift/pull/20005, so I'm going to go ahead and close it.
